### PR TITLE
Dev arrow orientation

### PR DESCRIPTION
**Pull request by roylly**
### Checklist for this pull request

- [x] I have read the [guidelines for contributing](../CONTRIBUTING.md)
- [x] Checked that my code additions did fail neither code linting checks nor unit/UI test.

### Description

 i have finished the arrow orientation support left and right development, but don't make big change for the layout engine, just add new case for layout. And the arrowView isn't a UIImageView anymore, but a UIView,and draw the arrow according the orientation, so the “highlighted” not work temporarily(i think that's not hard to fix). While , i also add backgroundImageView var for interface of BodyView, then user can hidden it,and set custom background color if they want.

(Thanks!)

### CODE REVIEW

**Comment by ephread** [comment]
Thanks again for the work.

I can't accept the PR as is, sadly. The code review that follows might be a bit overwhelming, but it's not meant to belittle your hard work in any way. I'll begin with the presentation, then I'll discuss the content.

Additionally, I will comment on specific lines of the diff.

### Presentation

To be fair, I didn't exactly write a [`CONTRIBUTING.md`](https://github.com/ephread/Instructions/blob/master/CONTRIBUTING.md) before, so that one is partially on me. There are general consistency issues. Here's the stuff which will need to be addressed:

1. **Linting** • The code does not pass Swiftlint's linting. You'll need to install Swiftlint and make sure to correct any warning/error you may encounter.

2. **Comments** • Some comments are written using Chinese characters, while I don't hold any particular view on which language should be used while coding, it needs to remain consistent throughout the code. For _Instructions_, comments have to be written in English.

3. **Commit messages** • While I'm fairly lenient with commit messages, they should still follow the standard rules. They should be no longer than 80 characters and if details are necessary, these can be put on the next line. Similarly, they shouldn't be written as the first person. Rather, they should describe the change in an impersonal way. Finally, commits should be written in English. This won't be too much of an issue however, as I will squash them before merging, but keep this in mind if you intend to submit other PRs.

### Content

- I'm not sure I understand why `CoachMarksViewController` and most of its fields became a public class. `CoachMarksViewController` has to remain internal, it's not supposed to be exposed as a public API. `CoachMarksController.coachMarksViewController` needs to remain private as well. The rational behind this choice, is that _Instructions_'s internals can radically change without breaking the API for everyone.

- Supplementary cutoutPaths (`extraViews`, `extraCutoutPath`) are included but don't seem to be used in any way. I'm not sure I understand where this is going. If multiple cutout paths were to be implemented, the delegate would need to support this as well; besides, I believe it wouldn't make much sense to keep both `cutoutPath` and `extraCutoutPaths`, as a single array could do the same. This very much seems to be an unrelated work in progress, so it shouldn't be included in this PR.

- While I'm not against using Core Graphics instead of images for the coach marks, it needs to be fully transparent for users. In this particular case, I see three issues with the current implementation:

    1. the highlight system no longer works on the arrow, I know that you said that you would fix it later, but it needs to work for the PR to be accepted;
    2. the arrow is missing its border; it needs to look like the body, it's about consistency;
    3. there's a one-pixel gap between the arrow and the body.

    All in all, I don't think it's necessary to use Core Graphics yet. While it's an entirely valid option for the future, I don't feel it provides a significant improvement. I've attached the proper graphical assets (left/right versions of the arrow) at the end of this comment, use these images and you won't need to rewrite the entire system.

### About background colors

I made the opinionated choice to not support background colors on default views (or any kind of customization beside hiding the `next` button). These views are meant to be an example. As I said on #155, if you need custom background colors, you should provide your own views, since I can't provide a one-size-fits-all solution. Besides, providing custom views is extremely easy.

What could happen in the future, however, is to have a subproject of _Instructions_, lying in another repository, providing a wide range of extensible, good looking, coach marks. Core Graphics based coach marks would be perfectly at home in such place and I believe it could satisfy everyone. ;)

[Arrows.zip](https://github.com/ephread/Instructions/files/2064549/Arrows.zip)

File: Examples/Example/Source/Core/View Controllers/ArrowOrientationLeftAndRightViewController.swift [hunk 1]
```diff
@@ -0,0 +1,76 @@
+//
```
Comment (ephread) [inline Examples/Example/Source/Core/View Controllers/ArrowOrientationLeftAndRightViewController.swift]: Here, you need to use the same MIT header as the other files, rather than the default one generated by Xcode.

File: Sources/Instructions/Core/Internal/CoachMarksViewController.swift [hunk 2]
```diff
@@ -27,7 +27,7 @@ import UIKit
 
 // MARK: - Main Class
 /// Handles a set of coach marks, and display them successively.
-class CoachMarksViewController: UIViewController {
+public class CoachMarksViewController: UIViewController {
```
Comment (ephread) [inline Sources/Instructions/Core/Internal/CoachMarksViewController.swift]: This class (and its fields) needs to remain `internal`.

File: Sources/Instructions/Core/Public/CoachMark.swift [hunk 3]
```diff
@@ -33,6 +33,7 @@ public struct CoachMark {
     /// The path to cut in the overlay, so the point of interest will be visible.
     public var cutoutPath: UIBezierPath?
 
+    public var extraCutoutPath : [UIBezierPath]?
```
Comment (ephread) [inline Sources/Instructions/Core/Public/CoachMark.swift]: Seems like a WIP, so the property and its related code found throughout the diff should not be included in this PR.

File: Sources/Instructions/Core/Public/CoachMarksController.swift [hunk 4]
```diff
@@ -81,7 +81,7 @@ public class CoachMarksController {
     fileprivate var coachMarksWindow: UIWindow?
 
     /// Handle the UI part
-    fileprivate lazy var coachMarksViewController: CoachMarksViewController = {
+    public lazy var coachMarksViewController: CoachMarksViewController = {
```
Comment (ephread) [inline Sources/Instructions/Core/Public/CoachMarksController.swift]: Needs to remain private as well.

File: Sources/Instructions/Extra/Default Views/CoachMarkArrowDefaultView.swift [hunk 5]
```diff
@@ -25,34 +25,88 @@ import UIKit
...
-        super.init(image: image, highlightedImage: highlightedImage)
+    }
+    public override init(frame: CGRect) {
+        super.init(frame: frame)
+    }
+    public init(orientation: CoachMarkArrowOrientation) {
+        self.orientation = orientation
+        super.init(frame: CGRect.zero)
+        self.backgroundColor = UIColor.clear
+//        let image, highlightedImage: UIImage?
```
Comment (ephread) [inline Sources/Instructions/Extra/Default Views/CoachMarkArrowDefaultView.swift]: Side note: you shouldn't include commented code in the PR ;)

File: Sources/Instructions/Extra/Default Views/CoachMarkArrowDefaultView.swift [hunk 6]
```diff
@@ -25,34 +25,88 @@ import UIKit
 // MARK: - Default Class
 /// A concrete implementation of the coach mark arrow view and the
 /// default one provided by the library.
-public class CoachMarkArrowDefaultView: UIImageView, CoachMarkArrowView {
+public class CoachMarkArrowDefaultView: UIView, CoachMarkArrowView {
```
Comment (ephread) [inline Sources/Instructions/Extra/Default Views/CoachMarkArrowDefaultView.swift]: The arrow should remain an `UIImageView` for now. See the main comment for details.

File: Sources/Instructions/Extra/Default Views/CoachMarkBodyDefaultView.swift [hunk 7]
```diff
@@ -46,6 +46,15 @@ public class CoachMarkBodyDefaultView: UIControl, CoachMarkBodyView {
 
     public var nextLabel: UILabel { return views.nextLabel }
     public var hintLabel: UITextView { return views.hintLabel }
+    public var backgroundImageView: UIImageView {
```
Comment (ephread) [inline Sources/Instructions/Extra/Default Views/CoachMarkBodyDefaultView.swift]: Defaults views are not meant to be customized, see main comment.

File: Sources/Instructions/Managers/Public/OverlayManager.swift [hunk 8]
```diff
@@ -24,7 +24,7 @@ import UIKit
 
 // Overlay a blocking view on top of the screen and handle the cutout path
 // around the point of interest.
-public class OverlayManager {
+public class OverlayManager:NSObject {
```
Comment (ephread) [inline Sources/Instructions/Managers/Public/OverlayManager.swift]: I don't see any valid reasons for OverlayManager to inherit NSObject. I can see that it was done in a80a085. From what I can understand from the commit message using automated translation, it was done to make it play well with swizzling in your project. While you certainly have good reasons for the change, I don't think it should be included in the base library.

File: Sources/Instructions/Views/CoachMarkView.swift [hunk 9]
```diff
@@ -25,7 +25,7 @@ import UIKit
 // swiftlint:disable force_cast
 
 /// The actual coach mark that will be displayed.
-class CoachMarkView: UIView {
+public class CoachMarkView: UIView {
```
Comment (ephread) [inline Sources/Instructions/Views/CoachMarkView.swift]: Again, it needs to remain internal.

File: Sources/Instructions/Views/CoachMarkView.swift [hunk 10]
```diff
@@ -133,21 +133,45 @@ class CoachMarkView: UIView {
...
+                        horizontalOffset: 0)
+                
+                    self.addConstraint(innerConstraints.arrowXposition!)
+                    //竖直方向的约束，按照箭头的方向，调整上下的位置
+                    self.addConstraints(coachMarkLayoutHelper.verticalConstraints(
+                        for: (bodyView: bodyUIView, arrowView: arrowUIView), in: self,
+                        withProperties: (orientation: arrowOrientation, verticalArrowOffset: arrowOffset)
+                    ))
+            case .left,.right:
+                print("test left and right")
```
Comment (ephread) [inline Sources/Instructions/Views/CoachMarkView.swift]: Debug prints like this one should be removed before submiting the PR.

File: Examples/Example/Source/Core/View Controllers/ArrowOrientationLeftAndRightViewController.swift (hunk 1 above)
Comment (roylly) [reply Examples/Example/Source/Core/View Controllers/ArrowOrientationLeftAndRightViewController.swift]: done

File: Sources/Instructions/Core/Internal/CoachMarksViewController.swift (hunk 2 above)
Comment (roylly) [reply Sources/Instructions/Core/Internal/CoachMarksViewController.swift]: I will add an example for this. As i want to get currentCoachMarkView to add a animation when it will or did show.

Comment (roylly) [reply Sources/Instructions/Core/Internal/CoachMarksViewController.swift]: do you have some advice for custom animation of CoachMarkView. I think what you said is right, CoachMarksViewController has to remain internal

File: Sources/Instructions/Core/Public/CoachMark.swift (hunk 3 above)
Comment (roylly) [reply Sources/Instructions/Core/Public/CoachMark.swift]: yes, the extraCutoutPath used when we need more than one highlighted area. sorry about that ,i will add example code and  make another PR for that .

File: Sources/Instructions/Core/Internal/CoachMarksViewController.swift (hunk 2 above)
Comment (ephread) [reply Sources/Instructions/Core/Internal/CoachMarksViewController.swift]: Adding animations to coach marks has to be done in another PR.

I'm currently drafting how it will work and I plan to use delegate methods as usual. I would use something in the spirit of `UIViewControllerTransitionCoordinator`, which would be provided to developers through the delegate methods. The developer could then pass the animation block to its `animate` method.

It could look like this:

```swift
func coachMarksController(_ coachMarksController: CoachMarksController, willAnimateAppearanceOfCoachMark view: UIView, at index: Int, using coordinator: CoachMarkTransitionCoordinator) {
    coordinator.animate {
        view.alpha = 0.0
    }
}
```

That said, I plan to work on this right after fixing the few open issues about the layout engine, so you're probably better off focusing on other aspects of Instructions. I need to be able to animate coach marks as well for an app that I'm making. ;)

**Comment by ephread** [comment]
I see that you have merged `master` back. I branched `layout-engine` from the same commit as you, so that you could work in total isolation, while I would continue to update master. Given that your changes are not simple fixes, I would have preferred handling merge conflicts myself, sorry if I didn't give you a heads up.

Additionally the PR now shows 157 files changed, most of them are noise and make the review process harder. Would you be able to revert the upstream merge in your branch and start fixing the PR from the last original commit (which was, if I'm not mistaken, 69145b2).

You could also create an entirely new PR, if you wish, by re-starting from `master`'s latest commit and re-adding your work on top.

Either way is fine by me. Sorry to be a bit stern, but I need the process to be clean. PR must have a relatively narrow scope and change a limited number of files (unless they are related to the project architecture). I'll clarify that in `CONTRIBUTING.md`.

**Comment by roylly** [comment]
i have checkout a new branch to make the PR, close this one! thanks.
